File: c/src/kdf.c
#include "kdf.h"

#include <argon2.h>
#include <openssl/evp.h>
#include <string.h>

#include "dorado/blake3.h"
#include "dorado/skein.h"

const char *dorado_kdf_derive(const dorado_kdf_params *p, const uint8_t *password, size_t password_len,
                              const uint8_t *salt, size_t salt_len, uint8_t *out, size_t out_len) {
    switch (p->kind) {
        case DORADO_KDF_ARGON2ID: {
            int rc = argon2id_hash_raw(p->t_cost, p->m_cost, p->p_cost, password, password_len, salt, salt_len, out,
                                       out_len);
            return rc == ARGON2_OK ? NULL : "argon2id derivation failed";
        }
        case DORADO_KDF_SCRYPT: {
            uint64_t n = (uint64_t)1 << p->log_n;
            /* Size maxmem to the parameters; this only gates the computation. */
            uint64_t maxmem = (uint64_t)128 * p->r * (n + p->p + 2) + (1u << 20);
            if (EVP_PBE_scrypt((const char *)password, password_len, salt, salt_len, n, p->r, p->p, maxmem, out,
                               out_len) != 1) {
                return "scrypt derivation failed";
            }
            return NULL;
        }
        case DORADO_KDF_PBKDF2: {
            if (PKCS5_PBKDF2_HMAC((const char *)password, (int)password_len, salt, (int)salt_len, (int)p->rounds,
                                  EVP_sha256(), (int)out_len, out) != 1) {
                return "pbkdf2 derivation failed";
            }
            return NULL;
        }
        default:
            return "unknown kdf kind";
    }
}

/* Fixed prefix domain-separating derive-from-key's keyed hashing from every
 * other keyed use in the engine ("DRDOrawE"/"DRDOrawM" in the raw-key split,
 * "DRDOchnk"/"DRDOrwFr" in the frame MACs). */
static const uint8_t DERIVE_FROM_KEY_DOMAIN[8] = {'D', 'R', 'D', 'O', 'k', 'd', 'r', 'v'};

const char *dorado_kdf_derive_from_key(const uint8_t *key, size_t key_len, const char *domain, uint8_t *out,
                                       size_t out_len) {
    return dorado_kdf_derive_from_key_with(DORADO_KDF_PRF_SKEIN512, key, key_len, domain, out, out_len);
}

const char *dorado_kdf_derive_from_key_with(int prf, const uint8_t *key, size_t key_len, const char *domain,
                                            uint8_t *out, size_t out_len) {
    /* out = PRF(key, out_len, "DRDOkdrv" || domain). Uses the port's own
     * from-scratch Skein-512/BLAKE3, not the delegated password-KDF libraries;
     * streaming update(A) then update(B) equals a one-shot MAC over A || B. */
    size_t domain_len = strlen(domain);
    switch (prf) {
        case DORADO_KDF_PRF_SKEIN512: {
            dorado_skein512 s;
            dorado_skein512_init_mac(&s, key, key_len, out_len);
            dorado_skein512_update(&s, DERIVE_FROM_KEY_DOMAIN, sizeof DERIVE_FROM_KEY_DOMAIN);
            dorado_skein512_update(&s, (const uint8_t *)domain, domain_len);
            dorado_skein512_finalize(&s, out);
            return NULL;
        }
        case DORADO_KDF_PRF_BLAKE3: {
            /* BLAKE3's keyed mode is defined only for a 256-bit key. */
            if (key_len != 32) {
                return dorado_err_params;
            }
            dorado_blake3 h;
            dorado_blake3_init_keyed(&h, key);
            dorado_blake3_update(&h, DERIVE_FROM_KEY_DOMAIN, sizeof DERIVE_FROM_KEY_DOMAIN);
            dorado_blake3_update(&h, (const uint8_t *)domain, domain_len);
            dorado_blake3_finalize(&h, out, out_len);
            return NULL;
        }
        default:
            return dorado_err_params;
    }
}

const char *dorado_kdf_validate(const dorado_kdf_params *p) {
    /* Out-of-range cost parameters are the params class (see engine.h). */
    switch (p->kind) {
        case DORADO_KDF_ARGON2ID:
            if (p->m_cost > (1u << 21) || p->t_cost > 64 || p->p_cost > 16) {
                return dorado_err_params;
            }
            return NULL;
        case DORADO_KDF_SCRYPT:
            if (p->log_n > 21 || p->r > 32 || p->p > 16) {
                return dorado_err_params;
            }
            return NULL;
        case DORADO_KDF_PBKDF2:
            /* Zero rounds would "derive" an all-zero key without error. */
            if (p->rounds == 0 || p->rounds > 50000000u) {
                return dorado_err_params;
            }
            return NULL;
        default:
            return dorado_err_params;
    }
}

File: zig/src/kdf.zig
//! Key derivation, in its two standard forms.
//!
//! `derive` is password-based derivation (a PBKDF): it stretches a weak,
//! guessable secret into a raw key, deliberately slowly, under caller-tunable
//! cost parameters (`validate` bounds untrusted ones). The three algorithms
//! (Argon2id, scrypt, PBKDF2-HMAC-SHA256) come from Zig's standard library
//! (`std.crypto.pwhash`); they are standard, so the derived keys match the
//! Rust reference byte-for-byte. Unlike the other ports, no external library
//! is needed.
//!
//! `deriveFromKey` is key-based derivation (a KBKDF): it splits an already
//! high-entropy key into independent, domain-separated children, fast (one
//! keyed hash), with no salt and no cost parameters because there is nothing
//! to stretch. The keyed hash defaults to Skein-512 (Threefish's native
//! companion); `deriveFromKeyWith` lets a caller pick the PRF (`KdfPrf`)
//! instead, e.g. BLAKE3 to keep a ChaCha-family construction single-family
//! top to bottom. Both use this port's own from-scratch primitives, not
//! `std.crypto`. The names are the guardrail: a password must never take the
//! fast path, and a key never needs the slow one.

const std = @import("std");
const argon2 = std.crypto.pwhash.argon2;
const scrypt = std.crypto.pwhash.scrypt;
const pbkdf2_fn = std.crypto.pwhash.pbkdf2;
const HmacSha256 = std.crypto.auth.hmac.sha2.HmacSha256;
const fmt = @import("format.zig");
const skein = @import("skein.zig");
const blake3 = @import("blake3.zig");

pub const Error = error{
    OutOfMemory,
    KdfFailed,
    HostileCost,
    BadKeyLength,
};

/// Stretch password (with salt) into out using params.
pub fn derive(
    allocator: std.mem.Allocator,
    io: std.Io,
    kdf: fmt.Kdf,
    password: []const u8,
    salt: []const u8,
    out: []u8,
) Error!void {
    switch (kdf) {
        .argon2id => |a| {
            argon2.kdf(
                allocator,
                out,
                password,
                salt,
                .{ .t = a.t_cost, .m = a.m_cost, .p = @intCast(a.p_cost) },
                .argon2id,
                io,
            ) catch return Error.KdfFailed;
        },
        .scrypt => |s| {
            scrypt.kdf(
                allocator,
                out,
                password,
                salt,
                .{ .ln = @intCast(s.log_n), .r = @intCast(s.r), .p = @intCast(s.p) },
            ) catch return Error.KdfFailed;
        },
        .pbkdf2 => |p| {
            pbkdf2_fn(out, password, salt, p.rounds, HmacSha256) catch return Error.KdfFailed;
        },
    }
}

/// The keyed hash `deriveFromKeyWith` fans a master key out with. Both are
/// secure PRFs and produce identically strong children; the choice exists only
/// to let a construction stay within one cryptographic family (Skein for
/// Threefish, BLAKE3 for a ChaCha-family cipher) rather than mixing lineages.
pub const KdfPrf = enum {
    /// Skein-512 keyed hash (Threefish's native companion). The default, and
    /// what `deriveFromKey` uses. Accepts a key of any length.
    skein512,
    /// BLAKE3 keyed hash. Requires a 32-byte key (BLAKE3's keyed mode is
    /// defined only for a 256-bit key); other lengths are `Error.BadKeyLength`.
    blake3,
};

/// Fixed prefix domain-separating `deriveFromKey`'s keyed hashing from every
/// other keyed use in the engine (`DRDOrawE`/`DRDOrawM` in the raw-key split,
/// `DRDOchnk`/`DRDOrwFr` in the frame MACs).
const DERIVE_FROM_KEY_DOMAIN = "DRDOkdrv";

/// Derive `out.len` key bytes from an already high-entropy `key`, separated by
/// `domain` -- key-based derivation (the fast form): one domain-separated
/// Skein-512 keyed hash, no salt, no cost parameters, because a strong key has
/// nothing to stretch. Deterministic: the same key and domain always yield the
/// same bytes, and different domains yield computationally unrelated ones, so a
/// caller can fan one master key out into independent per-purpose keys
/// (`deriveFromKey(master, "myapp/index", ..)`,
/// `deriveFromKey(master, "myapp/data", ..)`). Never pass a password here:
/// there is no stretching, so a guessable input stays guessable -- that is
/// `derive`'s job. To fan out with a different PRF (e.g. BLAKE3), use
/// `deriveFromKeyWith`.
pub fn deriveFromKey(key: []const u8, domain: []const u8, out: []u8) void {
    deriveFromKeyWith(.skein512, key, domain, out) catch unreachable;
}

/// `deriveFromKey` with a caller-chosen PRF (`KdfPrf`). The domain separation,
/// determinism, and "never pass a password" contract are exactly the same;
/// only the underlying keyed hash changes. With `.skein512` this is
/// byte-for-byte identical to `deriveFromKey`. `.blake3` requires `key` to be
/// 32 bytes and returns `Error.BadKeyLength` otherwise.
pub fn deriveFromKeyWith(prf: KdfPrf, key: []const u8, domain: []const u8, out: []u8) Error!void {
    switch (prf) {
        .skein512 => {
            // Streaming update(A) then update(B) equals update(A ++ B), so
            // this matches a one-shot MAC over the concatenated message.
            var h = skein.Skein512.initMac(key, out.len);
            h.update(DERIVE_FROM_KEY_DOMAIN);
            h.update(domain);
            h.finalize(out);
        },
        .blake3 => {
            if (key.len != 32) return Error.BadKeyLength;
            var h = blake3.Hasher.initKeyed(key[0..32]);
            h.update(DERIVE_FROM_KEY_DOMAIN);
            h.update(domain);
            h.finalize(out);
        },
    }
}

/// Reject KDF parameters whose cost is unreasonably large (the cost comes from
/// an untrusted header) or nonsensical (zero PBKDF2 rounds). Bounds match the
/// other ports.
pub fn validate(kdf: fmt.Kdf) Error!void {
    switch (kdf) {
        .argon2id => |a| {
            if (a.m_cost > (1 << 21)) return Error.HostileCost;
            if (a.t_cost > 64) return Error.HostileCost;
            if (a.p_cost > 16) return Error.HostileCost;
        },
        .scrypt => |s| {
            if (s.log_n > 21) return Error.HostileCost;
            if (s.r > 32) return Error.HostileCost;
            if (s.p > 16) return Error.HostileCost;
        },
        .pbkdf2 => |p| {
            // Zero rounds would "derive" an all-zero key without error.
            if (p.rounds == 0) return Error.HostileCost;
            if (p.rounds > 50_000_000) return Error.HostileCost;
        },
    }
}
